mysql-test/suite/binlog/t/binlog_recover_async_rollback_trx.test

```diff
@@ -0,0 +1,149 @@
+--let $assert_count= 1
+--source include/assert_grep.inc
+--remove_file $MYSQLTEST_VARDIR/tmp/log.err
+
+# The previous INSERT is rolled back, so this INSERT will succeed.
+INSERT INTO t1 VALUES(1, 1);
+
+--echo # Expected (0, 1), (1, 1)
+SELECT * FROM t1;
+
+--echo # Expect (0,2), (1,1)
+SET SESSION TRANSACTION ISOLATION LEVEL READ UNCOMMITTED;
+SELECT * FROM t1;
+
+--echo #
+--echo # 2. Check DML can rollback automatically after second crash.
+--echo #
+--source include/kill_mysqld.inc
+--let $restart_parameters=
+--source include/start_mysqld.inc
+--let $wait_condition= SELECT count(*) = 0 FROM information_schema.innodb_trx
+--source include/wait_condition.inc
+
+# Check whether the DML trx has been rolled back asynchronously.
+--let $assert_file= $MYSQLTEST_VARDIR/log/mysqld.1.err
+--let $assert_text= Server reported: Rolled back recovered transaction
+--let $assert_select= Rolled back recovered transaction
+--let $assert_count= 1
+--source include/assert_grep.inc
+
+# The transaction was set to ACTIVE in previous recovery, thus it should not
+# find any prepared thransactions.
+--let $assert_file= $MYSQLTEST_VARDIR/log/mysqld.1.err
+--let $assert_text= Server NOT report: Found 1 prepared transaction
+--let $assert_select= Found 1 prepared transaction
+--let $assert_count= 0
+--source include/assert_grep.inc
+
+UPDATE t1 SET data = data + 1 WHERE id = 0;
+--echo # expected (0, 2), (1, 1)
+SELECT * FROM t1;
+
+--echo # 3. Test that innodb shutdown as expected if any error happens before
+--echo #    normal rollback task is started. In the situation, rollback task
+--echo #    should be started at preshutdown accordingly to rollback or
+--echo #    deregister all recovered active transactions.
+# Generate an large active transaction
+--let $i = 2
+BEGIN;
+while ($i < 8)
+{
+  eval INSERT INTO t1 SELECT id + power(2, $i), 1 FROM t1;
+  --inc $i
+}
+COMMIT;
+
+BEGIN;
+UPDATE t1 SET data = 10;
+
+# Make sure above update is persisted.
+SET GLOBAL innodb_log_checkpoint_now = 1;
+--source include/kill_mysqld.inc
+
+# tc-heuristic-recover triggers an error before innodb rollback task start
+# Rollback task will not be started at preshutdown of read only mode. Active
+# transactions are not expected to rollback.
+--error 1
+--exec $MYSQLD_LAST_CMD --tc-heuristic-recover=ROLLBACK --innodb-read-only --log-error=$MYSQLTEST_VARDIR/tmp/log.err
+
+# Rollback task will not be started at preshutdown if recovery mode is greater
+# to 2.  Active transactions are not expected to rollback.
+--error 1
+--exec $MYSQLD_LAST_CMD --tc-heuristic-recover=ROLLBACK --innodb-force-recovery=3 --log-error=$MYSQLTEST_VARDIR/tmp/log.err
+
+# Rollback task will be started at preshutdown of fast shutdown if force
+# recovery is 2. But the transaction is deregistered instead of rollback.
+--error 1
+--exec $MYSQLD_LAST_CMD --tc-heuristic-recover=ROLLBACK --innodb-fast-shutdown=1 --innodb-force-recovery=2 --log-error=$MYSQLTEST_VARDIR/tmp/log1.err
```

Comment:
Why would it matter whether the transaction is fully rolled back? What should matter is that the transaction is no longer in `XA PREPARE` state and it is not causing blocking new transactions. (There could be a locking conflict until the transaction has been rolled back, but that conflict should be resolved automatically, unless `innodb_lock_wait_timeout` expires before the rollback completes.)

I think that the only way to observe the effect of the rollback would be to read data in a `READ UNCOMMITTED` transaction. Only with `innodb_fast_shutdown=0` there is a guarantee that all undo logs and transaction history will be cleared.

Reply:
The ROLLBACK option just helps to abort the server in this test, there is no PREPARED transaction in this test.  it doesn't rollback any active transactions.